### PR TITLE
Add hostPort support #144

### PR DESCRIPTION
**Pull request by igorvpcleao**
* Issue: #144 *

This PR adds support to hostPort.

### Design decisions to de discussed:
1. Not sure if hostPort support should be offered by `amazon-vpc-cni-k8s` image itself or by modifying amazon-vpc-cni manifests on Kops, adding a new image to perform a slightly different installation, for instance.
2. Portmap plugin is downloaded from `github.com/containernetworking/plugins/releases`

### CODE REVIEW

**Comment by cjbottaro** [comment]
I built and installed this in my cluster. Seem to work! 👍 Thank you very much for this!

**Comment by brockoffdev** [comment]
Would love to see this merged in, this is actually a pretty big blocker for our team on EKS 😄 👍

**Comment by liwenwu-amazon** [comment]
@igorvpcleao Thank you for submitting this PR.  
Can you add some detail steps:
* how to reproduce the issue
* how to verify this PR has fixed the issue

**Comment by igorvpcleao** [comment]
@liwenwu-amazon Sure. We need to create a pod and then set the `hostPort` field in its spec. After that, we make requests to `ip_from_node_where_pod_is_running:hostPort` and we should get an answer. 

Here you are a manifest for an app which replies `Hey. Look, I'm alive!`
`https://gist.github.com/igorvpcleao/a0f5087fa32f3438f3a0e912adf0a608`

If the node security group is open to where you are performing the request, via internet, you may run something like:
```
curl 54.209.x.x:51080
Hey. Look, I'm alive!%
```
Or you can ssh to a node from within you Kubernetes cluster and make the request:
```
curl 172.50.58.153:51080
Hey. Look, I'm alive!
```

Where `172.50.58.153` is a private ip for the node where your pod is running and `54.209.x.x` is its public ip.

**Comment by liwenwu-amazon** [comment]
@igorvpcleao Thanks for providing the test scenario.  I have one question:

After I did `kubectl apply -f manifests.yaml` , I can see following
```
# find out hey-iam-alive-hostport Pod IP
 kubectl get all -n hey-iam-alive-poc-awc-cni -o wide
NAME                     READY     STATUS    RESTARTS   AGE       IP                NODE
hey-iam-alive-hostport   1/1       Running   0          18m       192.168.254.207   ip-192-168-238-237.us-west-2.compute.internal
```
From one of the Pod running in the same cluster, I am able to get response from Pod using Pod's IP
```
# here are other Pods running in the same cluster
kubectl get pod -o wide
NAME                            READY     STATUS    RESTARTS   AGE       IP                NODE
hostport-nginx                  1/1       Running   0          40m       192.168.86.148    ip-192-168-125-35.us-west-2.compute.internal
nginx-65899c769f-9pttb          1/1       Running   0          46m       192.168.167.200   ip-192-168-178-137.us-west-2.compute.internal
worker-hello-5974f49799-22sjl   1/1       Running   0          45m       192.168.114.218   ip-192-168-125-35.us-west-2.compute.internal
worker-hello-5974f49799-5nkb4   1/1       Running   0          45m       192.168.212.76    ip-192-168-238-237.us-west-2.compute.internal
worker-hello-5974f49799-lj4kn   1/1       Running   0          45m       192.168.85.77     ip-192-168-125-35.us-west-2.compute.internal
worker-hello-5974f49799-ppvzj   1/1       Running   0          45m       192.168.184.168   ip-192-168-178-137.us-west-2.compute.internal
worker-hello-5974f49799-vfqc2   1/1       Running   0          45m       192.168.195.140   ip-192-168-238-237.us-west-2.compute.internal

# log into one of these Pods
kubectl exec -ti worker-hello-5974f49799-22sjl sh

#  try to get the content of hey-iam-alive-hostport Pod
wget 192.168.254.207:5000
Connecting to 192.168.254.207:5000 (192.168.254.207:5000)
index.html           100% |*******************************************************************************************************************************************************************|    21   0:00:00 ETA
/ # cat index.html
Hey. Look, I'm alive!/ # 

```

In summary, I am able to reach Pod using Pod's IP and its container port.

I am curious to know what's the use cases  you need to use hostport?   

thanks

**Comment by zapman449** [comment]
@liwenwu-amazon The use-case is Ingress.  For us it would be required to expose 80/443 on the instance to an end-user-facing NLB.

**Comment by cjbottaro** [comment]
Same use case here.

Our clients require a whitelisted IP address to connect to. So we run a tcp proxy on an instance with an AWS EIP that just forwards to an AWS ALB.

File: Makefile [hunk 1]
```diff
@@ -17,6 +17,15 @@
 VERSION ?= $(shell git describe --tags --always --dirty)
 LDFLAGS ?= -X main.version=$(VERSION)
 
+# Download portmap plugin
+download-portmap:
```
Comment (micahhausler) [inline Makefile]: We currently [bake CNI binaries in our AMI](https://github.com/awslabs/amazon-eks-ami/blob/9736e73/install-worker.sh#L78-L86), but we should really put this process in an init container (See #158).

Can you remove this for now?

File: scripts/install-aws.sh [hunk 2]
```diff
@@ -1,8 +1,9 @@
 #!/usr/bin/env bash
 echo "=====Starting installing AWS-CNI ========="
-sed -i s/__VETHPREFIX__/${AWS_VPC_K8S_CNI_VETHPREFIX:-"eni"}/g /app/aws.conf
+sed -i s/__VETHPREFIX__/${AWS_VPC_K8S_CNI_VETHPREFIX:-"eni"}/g /app/10-aws.conflist
 cp /app/aws-cni /host/opt/cni/bin/
+cp /app/portmap /host/opt/cni/bin/
```
Comment (micahhausler) [inline scripts/install-aws.sh]: Can you remove this?

File: scripts/dockerfiles/Dockerfile.release [hunk 3]
```diff
@@ -6,7 +6,9 @@ RUN yum update -y && \
 WORKDIR /app
 
 COPY aws-cni /app 
-COPY misc/aws.conf /app
+COPY misc/10-aws.conflist /app
+
+COPY portmap /app
```
Comment (micahhausler) [inline scripts/dockerfiles/Dockerfile.release]: Can you remove this too?

Comment (igorvpcleao) [reply scripts/dockerfiles/Dockerfile.release]: @micahhausler  Sure, I removed my last commit. It only handled portmap plugin. Let me know if you need further changes.

**Comment by igorvpcleao** [comment]
@liwenwu-amazon Sometimes it may be a little hard to handle requests from outside the cluster when we have stateful applications, mainly when using UDP. `hostPort` may help to simplify our architecture.

**Comment by igorvpcleao** [comment]
Hi @liwenwu-amazon,
Let me know if you need any further changes.

**Comment by igorvpcleao** [comment]
Hi @micahhausler,

Let me know if you need any further changes.

Best,

File: scripts/install-aws.sh [hunk 4]
```diff
@@ -1,8 +1,8 @@
 #!/usr/bin/env bash
 echo "=====Starting installing AWS-CNI ========="
-sed -i s/__VETHPREFIX__/${AWS_VPC_K8S_CNI_VETHPREFIX:-"eni"}/g /app/aws.conf
+sed -i s/__VETHPREFIX__/${AWS_VPC_K8S_CNI_VETHPREFIX:-"eni"}/g /app/10-aws.conflist
 cp /app/aws-cni /host/opt/cni/bin/
 cp /app/aws-cni-support.sh /host/opt/cni/bin/
-cp /app/aws.conf /host/etc/cni/net.d/
+cp /app/10-aws.conflist /host/etc/cni/net.d/
```
Comment (micahhausler) [inline scripts/install-aws.sh]: Can you remove `aws.conf` if it already exists? The file will already be there on existing systems where this is an upgrade.

```bash
if [[ -f /host/etc/cni/net.d/aws.conf ]]; then
    rm /host/etc/cni/net.d/aws.conf
fi
```

**Comment by adammw** [comment]
@micahhausler will this PR be compatible with the current CNI release or blocked behind the changes in #158? Lack of hostPort functionality is a blocker for us to migrating to this CNI plugin.

**Comment by liwenwu-amazon** [comment]
@igorvpcleao please also rebase your change.

File: scripts/install-aws.sh (hunk 4 above)
Comment (igorvpcleao) [reply scripts/install-aws.sh]: @micahhausler sure. I can submit it in a second.

**Comment by igorvpcleao** [comment]
@liwenwu-amazon done.

File: scripts/install-aws.sh (hunk 4 above)
Comment (igorvpcleao) [reply scripts/install-aws.sh]: @micahhausler done. Let me know if you want me to comment it somehow.

**Comment by liwenwu-amazon** [comment]
@igorvpcleao thanks!  Testing your PR now...

**Comment by liwenwu-amazon** [comment]
@igorvpcleao For some reason, with this PR, I am NOT able to create any Pod.   Let me try again with top of tree to see if this is my environment issue    Are you able to test your PR ?

**Comment by igorvpcleao** [comment]
@liwenwu-amazon sure. Let me try to reproduce it.

**Comment by igorvpcleao** [comment]
@liwenwu-amazon I think it's because in this PR we do not have the portmap plugin anymore, as it was moved to https://github.com/aws/amazon-vpc-cni-k8s/issues/158. I will try to create a new branch and merge both to see what happens.

**Comment by igorvpcleao** [comment]
@liwenwu-amazon Is the init container from #158 already available?

**Comment by liwenwu-amazon** [comment]
@igorvpcleao No, it is not available yet.

**Comment by liwenwu-amazon** [comment]
@igorvpcleao  can you revert the change back to your original commit, so host port will work just by this PR.  When we  working on #158 , we can do same logic for portmap plugin.  thanks

**Comment by igorvpcleao** [comment]
> @igorvpcleao can you revert the change back to your original commit, so host port will work just by this PR. When we working on #158 , we can do same logic for portmap plugin. thanks

@liwenwu-amazon  Sure! I've just added it again. We just need to be aligned with @micahhausler about it.
Any problem, let me know.

**Comment by liwenwu-amazon** [comment]
@igorvpcleao can you rebase your change and make change in `config/v1.2` , also squash them into 1 commit?  thanks

**Comment by igorvpcleao** [comment]
@liwenwu-amazon Sure! Sorry for my delay. What changes do you want to be made on `config/v1.2`? It seems ok for me.

**Comment by liwenwu-amazon** [comment]
@igorvpcleao   Sorry, you don't need to  change to config/v1.2. Can you rebase and squash all of your commits into 1 single commit?   Once you do that, I will kick off testing.  thanks

**Comment by igorvpcleao** [comment]
@liwenwu-amazon No problem. Yep, It's done on #1a50c88. Let me know if you have any problem.

**Comment by igorvpcleao** [comment]
@liwenwu-amazon No problem. Yep, It's done on #ee54500bc2c79c5731066d71286411a490b01b10. Let me know if you have any problem.

**Comment by liwenwu-amazon** [comment]
@igorvpcleao Thanks!  Starting testing it now. Once it passed conformance test, I will add it back to 1.2.  thanks
